File: app/src/main/java/com/example/weizhenbin/wangebug/modules/news/uis/NewsListFragment.java
package com.example.weizhenbin.wangebug.modules.news.uis;

import android.graphics.drawable.Drawable;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v4.widget.SwipeRefreshLayout;
import android.support.v7.widget.DividerItemDecoration;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.text.TextUtils;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import com.chad.library.adapter.base.BaseQuickAdapter;
import com.example.weizhenbin.wangebug.R;
import com.example.weizhenbin.wangebug.base.BaseFragment;
import com.example.weizhenbin.wangebug.base.DataResultAdapter;
import com.example.weizhenbin.wangebug.modules.news.adapters.NewsListAdapter;
import com.example.weizhenbin.wangebug.modules.news.controllers.NewController;
import com.example.weizhenbin.wangebug.modules.news.entity.YiYuanNewsBean;

import java.util.ArrayList;
import java.util.List;

/**
 * Created by weizhenbin on 2018/8/23.
 */

public class NewsListFragment extends BaseFragment {

    private SwipeRefreshLayout srlRefresh;
    private RecyclerView rvDataList;
    private NewsListAdapter listAdapter;
    int page=1;

    List<YiYuanNewsBean.ShowapiResBodyBean.PagebeanBean.ContentlistBean> contentlistBeen=new ArrayList<>();

    private String channelId;
    private String channelName;

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        if (savedInstanceState!=null){
            channelId=savedInstanceState.getString("channelId");
            channelName=savedInstanceState.getString("channelName");
        }
    }

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        View view=inflater.inflate(R.layout.fm_news_list,null);
        initViews(view);
        initData();
        initEvent();
        return view;
    }

    @Override
    public void onSaveInstanceState(@NonNull Bundle outState) {
        outState.putString("channelId",channelId);
        outState.putString("channelName",channelName);
        super.onSaveInstanceState(outState);
    }

    private void setChannelId(String channelId) {
        this.channelId = channelId;
    }

    private void setChannelName(String channelName) {
        this.channelName = channelName;
    }

    public static NewsListFragment getFragment(String channelId, String channelName){
        NewsListFragment newsListFragment=new NewsListFragment();
        newsListFragment.setChannelId(channelId);
        newsListFragment.setChannelName(channelName);
        return newsListFragment;
    }


    private void initData() {
        listAdapter=new NewsListAdapter(getContext(),contentlistBeen);
        rvDataList.setAdapter(listAdapter);
        rvDataList.setLayoutManager(new LinearLayoutManager(getContext(),LinearLayoutManager.VERTICAL,false));
        DividerItemDecoration itemDecoration=new DividerItemDecoration(getContext(),DividerItemDecoration.VERTICAL);
        Drawable drawable=getResources().getDrawable(R.drawable.divider_line);
        itemDecoration.setDrawable(drawable);
        rvDataList.addItemDecoration(itemDecoration);
        listAdapter.bindToRecyclerView(rvDataList);
        listAdapter.disableLoadMoreIfNotFullPage();
    }
    protected void initEvent() {
        srlRefresh.setOnRefreshListener(new SwipeRefreshLayout.OnRefreshListener() {
            @Override
            public void onRefresh() {
                page=1;
                getData();
            }
        });
        listAdapter.setOnLoadMoreListener(new BaseQuickAdapter.RequestLoadMoreListener() {
            @Override
            public void onLoadMoreRequested() {
                getData();
            }
        },rvDataList);
    }
    private void getData(){
        if (TextUtils.isEmpty(channelId)&&TextUtils.isEmpty(channelName)){
            return;
        }
        NewController.getNewsListData(page,channelId,new DataResultAdapter<YiYuanNewsBean>(){
            @Override
            public void onStart() {
                super.onStart();
                if (page==1) {
                    srlRefresh.setRefreshing(true);
                }
            }
            @Override
            public void onSuccess(YiYuanNewsBean yiYuanNewsBean) {
                super.onSuccess(yiYuanNewsBean);
                srlRefresh.setRefreshing(false);
                if (yiYuanNewsBean!=null&&yiYuanNewsBean.getShowapi_res_code()==0){
                    if (yiYuanNewsBean.getShowapi_res_body()!=null&&yiYuanNewsBean.getShowapi_res_body().getPagebean()!=null){
                        if (page==1){
                            contentlistBeen.clear();
                        }
                        contentlistBeen.addAll(yiYuanNewsBean.getShowapi_res_body().getPagebean().getContentlist());
                        if(page==1){
                            listAdapter.setNewData(contentlistBeen);
                        }else {
                            listAdapter.notifyDataSetChanged();
                        }
                        if (yiYuanNewsBean.getShowapi_res_body().getPagebean().isLastPage()){
                            listAdapter.loadMoreEnd();
                        }else {
                            listAdapter.loadMoreComplete();
                            page++;
                        }
                    }else {
                        listAdapter.loadMoreEnd();
                    }
                }
            }
        });
    }

    private void initViews(View view) {
        srlRefresh =  view.findViewById(R.id.srl_refresh);
        rvDataList = view.findViewById(R.id.rv_data_list);
    }

    @Override
    protected void loadData() {
        if (contentlistBeen.isEmpty()) {
            getData();
        }
    }

    @Override
    protected String getPageTitle() {
        if (TextUtils.isEmpty(channelName)){
            return "全部";
        }
        return channelName;
    }
}

File: app/src/main/java/com/example/weizhenbin/wangebug/modules/news/uis/NewsListFragment.kt
package com.example.weizhenbin.wangebug.modules.news.uis

import android.os.Bundle
import android.support.v4.widget.SwipeRefreshLayout
import android.support.v7.widget.DividerItemDecoration
import android.support.v7.widget.LinearLayoutManager
import android.support.v7.widget.RecyclerView
import android.text.TextUtils
import android.view.View
import com.example.weizhenbin.wangebug.R
import com.example.weizhenbin.wangebug.base.BaseFragment
import com.example.weizhenbin.wangebug.base.DataResultAdapter
import com.example.weizhenbin.wangebug.modules.news.adapters.NewsListAdapter
import com.example.weizhenbin.wangebug.modules.news.controllers.NewController
import com.example.weizhenbin.wangebug.modules.news.entity.YiYuanNewsBean
import java.util.*

/**
 * Created by weizhenbin on 2018/8/23.
 */

class NewsListFragment : BaseFragment() {

    private var srlRefresh: SwipeRefreshLayout? = null
    private var rvDataList: RecyclerView? = null
    private var listAdapter: NewsListAdapter? = null
    internal var page = 1

    internal var contentListBeen: MutableList<YiYuanNewsBean.ShowapiResBodyBean.PagebeanBean.ContentlistBean> = ArrayList()

    private var channelId: String? = null
    private var channelName: String? = null


    override val contentViewLayoutId: Int
        get() = R.layout.fm_news_list

    override var pageTitle: String?
        get() = if (TextUtils.isEmpty(channelName)) {
            "全部"
        } else channelName
        set(value) {
            super.pageTitle = value
        }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        savedInstanceState?.let {
            channelId = it.getString("channelId")
            channelName = it.getString("channelName")
        }

    }

    override fun initFragment(view: View?) {
        initViews(view!!)
        initData()
        initEvent()
    }

    override fun onSaveInstanceState(outState: Bundle) {
        outState.putString("channelId", channelId)
        outState.putString("channelName", channelName)
        super.onSaveInstanceState(outState)
    }

    private fun setChannelId(channelId: String?) {
        this.channelId = channelId
    }

    private fun setChannelName(channelName: String?) {
        this.channelName = channelName
    }


    private fun initData() {
        if (context == null) {
            return
        }
        listAdapter = NewsListAdapter(context!!, contentListBeen)
        rvDataList!!.adapter = listAdapter
        rvDataList!!.layoutManager = LinearLayoutManager(context, LinearLayoutManager.VERTICAL, false)
        val itemDecoration = DividerItemDecoration(context!!, DividerItemDecoration.VERTICAL)
        val drawable = resources.getDrawable(R.drawable.divider_line)
        itemDecoration.setDrawable(drawable)
        rvDataList!!.addItemDecoration(itemDecoration)
        listAdapter!!.bindToRecyclerView(rvDataList)
        listAdapter!!.disableLoadMoreIfNotFullPage()
    }

    protected fun initEvent() {
        srlRefresh!!.setOnRefreshListener {
            page = 1
            getData()
        }
        listAdapter!!.setOnLoadMoreListener({ getData() }, rvDataList)

        listAdapter!!.setAction(View.OnClickListener {
            page = 1
            getData()
        })
    }

    private fun getData() {
        if (TextUtils.isEmpty(channelId) && TextUtils.isEmpty(channelName)) {
            return
        }
        NewController.getNewsListData(page, channelId!!, object : DataResultAdapter<YiYuanNewsBean>() {
            override fun onSuccess(yiYuanNewsBean: YiYuanNewsBean?) {
                super.onSuccess(yiYuanNewsBean)
                srlRefresh!!.isRefreshing = false
                if (yiYuanNewsBean != null && yiYuanNewsBean.showapi_res_code == 0) {
                    if (yiYuanNewsBean.showapi_res_body != null && yiYuanNewsBean.showapi_res_body!!.pagebean != null) {
                        if (page == 1) {
                            contentListBeen.clear()
                        }
                        contentListBeen.addAll(yiYuanNewsBean.showapi_res_body!!.pagebean!!.contentlist!!)
                        if (page == 1) {
                            listAdapter!!.setNewData(contentListBeen)
                        } else {
                            listAdapter!!.notifyDataSetChanged()
                        }
                        if (yiYuanNewsBean.showapi_res_body!!.pagebean!!.isLastPage) {
                            listAdapter!!.loadMoreEnd()
                        } else {
                            listAdapter!!.loadMoreComplete()
                            page++
                        }
                    } else {
                        listAdapter!!.loadMoreEnd()
                    }
                } else {
                    if (page == 1) {
                        listAdapter!!.emptyData()
                    } else {
                        listAdapter!!.loadMoreFail()
                    }
                }
            }

            override fun onError(throwable: Throwable) {
                super.onError(throwable)
                srlRefresh!!.isRefreshing = false
                if (page == 1) {
                    listAdapter!!.emptyData()
                } else {
                    listAdapter!!.loadMoreFail()
                }
            }
        })
    }

    private fun initViews(view: View) {
        srlRefresh = view.findViewById(R.id.srl_refresh)
        rvDataList = view.findViewById(R.id.rv_data_list)
    }

    override fun loadData() {
        if (contentListBeen.isEmpty()) {
            getData()
        }
    }

    companion object {

        fun getFragment(channelId: String?, channelName: String?): NewsListFragment {
            val newsListFragment = NewsListFragment()
            newsListFragment.setChannelId(channelId)
            newsListFragment.setChannelName(channelName)
            return newsListFragment
        }
    }
}
